aptos-move/move-examples/marketplace/sources/listing.move

```diff
@@ -0,0 +1,259 @@
+
+    use aptos_token::token::{Self as tokenv1, Token as TokenV1};
+    use aptos_token_objects::token;
+    use aptos_token_objects::royalty;
+
+    use marketplace::fee_schedule::FeeSchedule;
+
+    friend marketplace::coin_listing;
+
+    /// There exists no listing.
+    const ENO_LISTING: u64 = 1;
+    /// The listing is not yet live.
+    const ELISTING_NOT_STARTED: u64 = 2;
+    /// The entity is not the creator.
+    const ENOT_CREATOR: u64 = 3;
+
+    // Core data structures
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Corner-stone for all listings, represents the core utility layer including object
+    /// cleanup.
+    struct Listing has key {
+        /// The item owned by this listing, transferred to the new owner at the end.
+        object: Object<ObjectCore>,
+        /// The seller of the object
+        seller: address,
+        /// The fees associated with claiming this listing.
+        fee_schedule: Object<FeeSchedule>,
+        /// The Unix timestamp in seconds at which point bidding and purchasing can occur
+        start_time: u64,
+        /// Used to clean-up at the end.
+        delete_ref: DeleteRef,
+        /// Used to create a signer to transfer the listed item, ideally the TransferRef would
+        /// support this.
+        extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Contains a tokenv1 as an object
+    struct TokenV1Container has key {
+        /// The stored token.
+        token: TokenV1,
+        /// Used to cleanup the object at the end
+        delete_ref: DeleteRef,
+        /// Used to transfer the tokenv1 at the conclusion of a purchase.
+        transfer_ref: TransferRef,
+    }
+
+    // Init functions
+
+    public(friend) fun init(
+        creator: &signer,
+        object: Object<ObjectCore>,
+        fee_schedule: Object<FeeSchedule>,
+        start_time: u64,
+    ): (signer, ConstructorRef) {
+        let constructor_ref = object::create_object_from_account(creator);
+        // Once we construct this, both the listing and its contents are soulbound until the conclusion.
+        let transfer_ref = object::generate_transfer_ref(&constructor_ref);
+        object::disable_ungated_transfer(&transfer_ref);
+        let listing_signer = object::generate_signer(&constructor_ref);
+
+        let listing = Listing {
+            object,
+            seller: signer::address_of(creator),
+            fee_schedule,
+            start_time,
+            delete_ref: object::generate_delete_ref(&constructor_ref),
+            extend_ref: object::generate_extend_ref(&constructor_ref),
+        };
+        move_to(&listing_signer, listing);
+
+        let listing_addr = object::address_from_constructor_ref(&constructor_ref);
+        object::transfer(creator, object, listing_addr);
+
+        (listing_signer, constructor_ref)
+    }
+
+    public(friend) fun create_tokenv1_container(
```

Comment:
IMO it's better to just separate this out into a wrapping example (e.g. Greg has one here https://github.com/aptos-labs/aptos-core/pull/8032)

Reply:
This is different. This makes the assumption that we want the tokenv1 in the tokenv2. This in turn also creates an undiscoverable collection... that needs more work and isn't compatible with the goals of this project. Let me remove the tokenwarapper into its own file though.